File: CarHomeWidget/src/main/java/com/anod/car/home/prefs/SkinPreviewFragment.java
package com.anod.car.home.prefs;

import android.app.Activity;
import android.app.Fragment;
import android.app.LoaderManager;
import android.content.AsyncTaskLoader;
import android.content.Loader;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RemoteViews;

import com.anod.car.home.R;
import com.anod.car.home.appwidget.WidgetViewBuilder;
import com.anod.car.home.prefs.LookAndFeelActivity.SkinRefreshListener;

import butterknife.ButterKnife;
import butterknife.InjectView;

public class SkinPreviewFragment extends Fragment implements LoaderManager.LoaderCallbacks<View>, SkinRefreshListener {
	
	private static final String ARG_POSITION = "position";
	private int mPosition;
	private LookAndFeelActivity mActivity;
	@InjectView(R.id.container_preview) ViewGroup mContainer;

	public static SkinPreviewFragment newInstance(int position) {
		SkinPreviewFragment f = new SkinPreviewFragment();
		
		Bundle args = new Bundle();
		args.putInt(ARG_POSITION, position);

		f.setArguments(args);
		
		return f;
	}
	/* (non-Javadoc)
	 * @see android.support.v4.app.Fragment#onResume()
	 */
	@Override
	public void onResume() {
		super.onResume();
		getLoaderManager().initLoader(0, null, this).forceLoad();
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);
		mActivity.onPreviewStart(mPosition);	
		getLoaderManager().initLoader(0, null, this);
	}

	@Override
	public void onAttach(Activity activity) {
		super.onAttach(activity);
		mPosition = getArguments().getInt(ARG_POSITION);
		mActivity = (LookAndFeelActivity)activity;
		mActivity.onFragmentAttach(this, mPosition);
	}

	@Override
	public void onDetach() {
		mActivity.onFragmentDetach(mPosition);
		super.onDetach();
	}
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.skin_item, container, false);
        ButterKnife.inject(this,view);
		return view;
	}
	
	@Override
	public void refresh() {
		getLoaderManager().initLoader(0, null, this).forceLoad();
	}


	public static class ViewLoader extends AsyncTaskLoader<View> {
		private final LookAndFeelActivity mActivity;
		private final int mPosition;
		
		public ViewLoader(LookAndFeelActivity activity, int position) {
			super(activity);
			mActivity = activity;
			mPosition = position;
		}

		@Override
		public View loadInBackground() {
			WidgetViewBuilder builder = mActivity.createBuilder();
			builder.init();

			builder.setOverrideSkin(mActivity.getSkinItem(mPosition).value);
			RemoteViews rv = builder.build();

			return rv.apply( mActivity, null );
		}

	}
	
	@Override
	public Loader<View> onCreateLoader(int id, Bundle args) {
		return new ViewLoader(mActivity, mPosition);
	}

	@Override
	public void onLoadFinished(Loader<View> loader, View inflatedView) {
    	mActivity.onPreviewCreated(mPosition);
        if( mContainer.getChildCount() > 0 ) {
        	mContainer.removeAllViews();
        }
        // TODO
        if (inflatedView.getParent() != null) {
        	((ViewGroup)inflatedView.getParent()).removeView(inflatedView);
        }
        mContainer.addView( inflatedView );
		mContainer.invalidate();
        //mContainer.requestLayout();
	}

	@Override
	public void onLoaderReset(Loader<View> loader) {
		// TODO Auto-generated method stub
		
	}

}

File: app/src/main/java/com/anod/car/home/prefs/lookandfeel/SkinPreviewFragment.kt
package com.anod.car.home.prefs.lookandfeel

import android.content.ClipData
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import com.anod.car.home.R
import com.anod.car.home.appwidget.WidgetViewBuilder
import com.anod.car.home.prefs.LookAndFeelActivity
import com.anod.car.home.prefs.drag.ShortcutShadowBuilder
import info.anodsplace.applog.AppLog
import info.anodsplace.carwidget.content.model.WidgetShortcutsModel
import info.anodsplace.carwidget.preferences.DefaultsResourceProvider
import info.anodsplace.carwidget.screens.widget.SkinPreviewViewModel
import kotlinx.coroutines.flow.collect

class SkinPreviewFragment : Fragment(), View.OnLongClickListener {

    private var position: Int = 0

    private var lookAndFeelActivity: LookAndFeelActivity? = null

    private var viewGroup: ViewGroup? = null

    private var shortcutsCount: Int = 0

    private val viewModel: SkinPreviewViewModel by viewModels()

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        //lookAndFeelActivity!!.onPreviewStart(position)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        position = requireArguments().getInt(ARG_POSITION)
        lookAndFeelActivity = context as LookAndFeelActivity?
    }

    override fun onDetach() {
        super.onDetach()
        lookAndFeelActivity = null
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.skin_item, container, false)
        viewGroup = view.findViewById(R.id.container_preview)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //viewModel.appWidgetId = lookAndFeelActivity!!.appWidgetId
        load()
    }

    fun refresh() {
        load()
    }

    private fun load() {
        viewLifecycleOwner.lifecycleScope.launchWhenResumed {
//            val overrideSkin = lookAndFeelActivity!!.getSkinItem(position)
//            viewModel.load(overrideSkin).collect { inflatedView ->
//                lookAndFeelActivity!!.onPreviewCreated(position)
//
//                if (inflatedView.parent != null) {
//                    return@collect
//                }
//
//                if (viewGroup!!.childCount > 0) {
//                    viewGroup!!.removeAllViews()
//                }
//
//                val model = WidgetShortcutsModel.init(requireContext(), DefaultsResourceProvider(requireContext()), lookAndFeelActivity!!.appWidgetId)
//                shortcutsCount = model.count
//
//                setupDragNDrop(inflatedView, model)
//                viewGroup!!.addView(inflatedView)
//            }
        }
    }

    override fun onDestroyView() {
        if (viewGroup != null && viewGroup!!.childCount > 0) {
            for (pos in 0 until shortcutsCount) {
                val btnResId = WidgetViewBuilder.btnIds[pos]
                val dragButton = viewGroup!!.findViewById<View>(btnResId) as? ImageView
                if (dragButton == null) {
                    AppLog.e("Count: $shortcutsCount, pos: $pos")
                    continue
                }
                dragButton.setOnLongClickListener(null)
                dragButton.setOnDragListener(null)
            }
        }

        super.onDestroyView()
    }

    private fun setupDragNDrop(inflatedView: View, model: WidgetShortcutsModel) {

        for (pos in 0 until shortcutsCount) {
            val btnResId = WidgetViewBuilder.btnIds[pos]
            val btn = inflatedView.findViewById<View>(btnResId) as? ImageView
            if (btn == null) {
                AppLog.e("Count: $shortcutsCount, pos: $pos")
                continue
            }
            val shortcut = model.get(pos)

            initDragButton(pos, btn, shortcut != null)
        }
    }

    override fun onLongClick(dragButton: View): Boolean {
//        val dragData = "" + dragButton.tag
//        val data = ClipData.newPlainText(dragData, dragData)
//        lookAndFeelActivity!!.onBeforeDragStart()
//        dragButton.startDragAndDrop(data, ShortcutShadowBuilder(dragButton), null, 0)
        return true
    }

    private fun initDragButton(cellId: Int, dragButton: ImageView,
                               hasShortcut: Boolean) {
//        dragButton.tag = cellId.toString()
//        if (hasShortcut) {
//            dragButton.setOnLongClickListener(this)
//        }
//        dragButton.setOnDragListener(lookAndFeelActivity!!.dragListener)
    }

    companion object {
        private const val ARG_POSITION = "position"
        fun newInstance(position: Int) = SkinPreviewFragment().apply {
            arguments = bundleOf(ARG_POSITION to position)
        }
    }

}
